### DISCUSSION TITLE
多脚本支持/ Support multiple userscripts

### DISCUSSION THREAD

**Discussion by northsea4** [lisonge/vite-plugin-monkey#40] [Ideas]
目前`vite-plugin-monkey`的体验已经很不错了，感谢作者的付出👍🏻。

突发“奇”想，有了以下这些想法😁

## 需求
- 一个脚本，需要使用多种打包方式
  - 示例1：cdn资源版，即引用的资源都通过cdn加载
  - 示例2：本地资源版，即引用的资源都打包到脚本里

- 多个脚本，比如脚本的开发、打包等，可以在一个项目中进行

## 好处
1. 一个项目
2. 代码复用
3. 集中管理开发和打包等

## 期望的功能
1. 支持启动多个脚本的开发服务

2. 支持打包指定名称的脚本

3. 支持打包全部脚本

4. 支持命令行生成新脚本基础目录文件结构

**Comment by lisonge** [comment]
> 一个脚本，需要使用多种打包方式

这个可以通过命令传递参数判断打包方式然后在 package.json.scripts 传递参数
或者使用不同的 vite.config.xxx.ts 然后使用 `vite --config vite.config.xxx.ts` 即可

---

> 多个脚本，比如脚本的开发、打包等，可以在一个项目中进行

单个项目开发多个脚本，这个属于项目结构管理的问题，这超出了 vite/vite-plugin 功能范围

实际上 `monorepo`  非常适合你描述的场景, 它有多种实现，[pnpm monorepo](https://pnpm.io/zh/workspaces) 是广泛被使用的一种

比如下列仓库都在使用 `pnpm monorepo` 

- [userscripts](https://github.com/lisonge/userscripts)
- [vite-plugin-monkey](https://github.com/lisonge/vite-plugin-monkey)
- [vite](https://github.com/vitejs/vite)
- [vue](https://github.com/vuejs/vue)
- [element-plus](https://github.com/element-plus/element-plus)

拿 [userscripts](https://github.com/lisonge/userscripts) 举例

对于你说的好处

- [x] 一个项目
- [x] 代码复用，见 [shared/userscript-config](https://github.com/lisonge/userscripts/tree/main/shared/userscript-config)
- [x] 集中管理开发和打包等

对于你期望的功能

- [x] 支持启动多个脚本的开发服务，单个使用 `pnpm -F v2ex-comment-markdown serve`，多个使用`pnpm -r --filter=./packages/* serve`
- [x] 支持打包指定名称的脚本，使用 `pnpm -F v2ex-comment-markdown build`
- [x] 支持打包全部脚本，使用`pnpm -r --filter=./packages/* build`
- [x] 支持命令行生成新脚本基础目录文件结构，在packages目录使用`pnpm create monkey`然后手动删除公共依赖

最后一个功能稍微有点麻烦，因此 [create-monkey](https://github.com/lisonge/vite-plugin-monkey/tree/main/packages/create-monkey) 后续也会增加一个 template-monorepo

**Reply by northsea4** [reply]
感谢指点！
